### PR TITLE
[ao][sparsity] Support for sparsifying data operations on raw torch tensors.

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79535
* #79534
* #79254
* #79253
* #79427
* __->__ #79252
* #79251

The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Differential Revision: [D37164121](https://our.internmc.facebook.com/intern/diff/D37164121)

### CODE REVIEW

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 1]
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
...
+        format_string += ')'
+        return format_string
 
     def get_mask(self, name: str):
-        pass
+        if name not in self.state:
+            raise ValueError("data with specified name does not exist")
+        return self.state[name]['mask']
 
     def squash_mask(self, *args, **kwargs):
```
Comment (HDCharles) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:226]: should probably leave the kwargs in the function definition rather than just using kwargs, 

I believe if you do something like def squash_mask(self, *args, leave_parametrize=True, **kwargs) it will work.

**Review by HDCharles — APPROVED** [review]
looks good aside from those comments, not going to block though

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 2]
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
...
     def add_data(self, name: str, data, **config):
         r""" Configures and parametrizes the internal container model with name and data
 
         Note: The container model is private to the BaseDataSparsifier class
         """
-        pass
+        assert type(data) in SUPPORTED_TYPES, \
+            f'data type unsupported for {name}'
+
+        if name in self.state:
```
[the remark this answers is not in the document]
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:71]: Yep - can do. Will make changes.

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 1 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:226]: got it

**Comment by z-a-f** [comment]
More descriptive name for the PR title is needed. `Support for toech tensors` doesn't tell a lot about where is this support added

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 3]
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
...
+        self.__dict__.update(state)
 
     def __getstate__(self):
-        pass
+        return {
+            'defaults': self.defaults,
+            'state': self.state,
+            'data_groups': self.data_groups,
+            '_container': self._container.state_dict()
+        }
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:206]: Let's split out the serialization routines into a separate diff -- there might be more focusd comments that way, and there might be some tests that we would apply

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 4]
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
...
+        if isinstance(data, torch.Tensor):
+            return data
+
     def add_data(self, name: str, data, **config):
         r""" Configures and parametrizes the internal container model with name and data
 
         Note: The container model is private to the BaseDataSparsifier class
         """
-        pass
+        assert type(data) in SUPPORTED_TYPES, \
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:68]: nit: `assert isinstance(data, SUPPORTED_TYPES)`

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 5]
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
...
+                else - returns the original version (non-parametrized)
+        """
+        if name not in self.data_groups:
+            raise ValueError("data with specified name does not exist")
 
-    def get_data(self, name: str):
-        pass
+        if not return_sparsified:
+            if not parametrize.is_parametrized(self._container, name):
+                raise ValueError("mask squashed - original mask value does not exist")
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:110]: nit: If it's not parametrized, and the assumption is that the squash_mask was already called, shouldn't this just return the tensor from the container? If the intention is to return the parametrized vs. original, I think a better name for the argument shoud be something like

```
def get_data(self, name: str, return_original: bool = True):
```

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py [hunk 6]
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
...
+            states.update(self.state)
+            data_groups.update(self.data_groups)
+
+        self.__setstate__({'state': states, 'data_groups': data_groups})
 
     def __setstate__(self, state):
-        pass
+        if '_container' in state:  # If container object is in state then load model
+            container_dict = state.pop('_container')
+            self._container = Container()
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:195]: Is there a reason why a new instance of the container is created, instead of applying the state to the existing container?

**Review by z-a-f — CHANGES_REQUESTED** [review]

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 3 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:206]: Oh I thought I did. I seemed to have amended the commit by mistake. Let me create a separate diff.

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 6 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:195]: The reason is that to do a load_state_dict() on an nn module, it is essential that the model parameters already exist in the container. Also, while unpickling (or deserializing) I don't think the constructor is called - hence we will have to create a new container object.

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 5 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:110]: makes sense.

File: test/ao/sparsity/test_data_sparsifier.py [hunk 7]
```diff
@@ -2,12 +2,152 @@
...
+        super().__init__(**kwargs)
+
+    def update_mask(self, name, data, **kwargs):
+        mask = self.get_mask(name)
+        mask[0] = 0
+        linear_state = self.state[name]
+        linear_state['step_count'] = linear_state.get('step_count', 0) + 1
+
+
+class TestBaseDataSparsiferType(TestCase):
```
Comment (HDCharles) [inline test/ao/sparsity/test_data_sparsifier.py:26]: I only just now understood whats going on with this test class, I hadn't realized before that this was a helper class for the actual test class TestBaseDataSparsifier. I'm not understanding the name here, feels like it should be something like TestBaseDataSparsifier(Generator/Helper/Runner) or something like that and it should have a comment similar to the comment for the base class describing this.

**Review by z-a-f — APPROVED** [review]
lgtm

File: torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py (hunk 3 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py]: done

File: test/ao/sparsity/test_data_sparsifier.py (hunk 7 above)
Comment (macandro96) [reply test/ao/sparsity/test_data_sparsifier.py]: The helper class could take in data of any time and then run tests on them, so I had suffixed with Type. Renamed it to TestBaseDataSparsifierRunner as it seems to be a better name. Added comments as well.

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164121).

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164121).

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164121).
